File: Backend/Application/DTO/Budget/Months/Editor/Savings/BudgetMonthSavingsGoalArchiveRowDto.cs
namespace Backend.Application.DTO.Budget.Months.Editor.Savings;

// Read-only projection of a closed (completed or cancelled) savings goal as
// surfaced by the "previous goals" archive on the savings editor. The active
// editor list (`BudgetMonthSavingsGoalEditorRowDto`) stays active-only — this
// DTO carries the closed history separately so the UI never has to filter the
// editor contract by status.
//
// `AmountSavedAtClose` is the canonical display value for the archive — it is
// derived server-side, never raw stored progression. Per the close-month
// contract, close-month does NOT advance AmountSaved, so a completed goal's
// raw `AmountSaved` will under-report the amount the user actually reached.
// Rule: completed → AmountSaved + MonthlyContribution; cancelled →
// AmountSaved. The UI must not recompute this.
public sealed record BudgetMonthSavingsGoalArchiveRowDto(
    Guid Id,
    Guid? SourceSavingsGoalId,
    string Name,
    decimal? TargetAmount,
    DateTime? TargetDate,
    decimal? AmountSavedAtClose,
    decimal MonthlyContribution,
    string Status,
    string ClosedReason,
    DateTime? ClosedAt,
    bool IsMonthOnly);

File: Frontend/src/types/budget/BudgetMonthSavingsGoalArchiveRowDto.ts
// Read-only projection of a closed savings goal as surfaced by the savings
// editor "previous goals" archive. The backend filters out removed and
// deleted rows server-side — the FE never has to skip them. Budget-scoped,
// so a goal completed in an earlier month still appears in later open
// months.
//
// `amountSavedAtClose` is the canonical display amount and is derived
// server-side: for `completed` goals it includes the closing-month
// contribution (close-month does NOT advance raw AmountSaved), for
// `cancelled` goals it equals what the user actually saved. The UI must
// not recompute this.
export type BudgetMonthSavingsGoalArchiveRowDto = {
  id: string;
  sourceSavingsGoalId: string | null;
  name: string;
  targetAmount: number | null;
  targetDate: string | null;
  amountSavedAtClose: number | null;
  monthlyContribution: number;
  status: string;
  closedReason: "completed" | "cancelled" | string;
  closedAt: string | null;
  isMonthOnly: boolean;
};
